craft_parts/state_manager/state_manager.py

```diff
@@ -0,0 +1,152 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Part crafter step state management."""
+
+import itertools
+from dataclasses import dataclass
+from typing import Dict, Optional
+
+from craft_parts.steps import Step
+
+from .states import StepState
+
+
+@dataclass(frozen=True)
+class _StateWrapper:
```

Comment:
Are these types going to be exposed outside of `state_manager` package?

Reply:
No, they're strictly internal.